File: apps/api/src/models/map.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BoundingBox {
    pub north: f64,
    pub south: f64,
    pub east: f64,
    pub west: f64,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MapClustersRequest {
    pub bounds: BoundingBox,
    pub zoom: u8,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Cluster {
    pub id: String,
    pub lat: f64,
    pub lng: f64,
    pub count: i64,
    pub representative_id: i64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MapClustersResponse {
    pub clusters: Vec<Cluster>,
    pub total_count: i64,
}

File: apps/web/src/api/map.ts
import { apiClient } from './client'
import type { Media } from './types'

export interface BoundingBox {
  north: number
  south: number
  east: number
  west: number
}

export interface MapClusterRequest {
  bounds: BoundingBox
  zoom: number
  filters?: Record<string, unknown>
}

export interface MapMediaRequest {
  bounds: BoundingBox
  geohashPrefixes?: string[]
}

export interface Cluster {
  id: string
  lat: number
  lng: number
  count: number
  representativeId: number
}

export interface MapClustersResponse {
  clusters: Cluster[]
  totalCount: number
}

export interface MapMediaResponse {
  items: Media[]
}

export const mapApi = {
  getClusters: async (bounds: BoundingBox, zoom: number): Promise<MapClustersResponse> => {
    const response = await apiClient.post<MapClustersResponse>('/map/clusters', {
      bounds,
      zoom,
    })
    return response.data
  },
  getMedia: async (request: MapMediaRequest): Promise<MapMediaResponse> => {
    const response = await apiClient.post<MapMediaResponse>('/map/media', request)
    return response.data
  },
}
